SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp

```diff
@@ -1842,52 +1865,105 @@ void USpatialNetDriver::ProcessRPC(AActor* Actor, UObject* SubObject, UFunction*
 
-			if (GSenderStack.Num() == 0)
+			if (bIsUnordered)
 			{
-				// These will eventually become error cases
-				// UE_LOG(LogSpatialOSNetDriver, Error, TEXT("Missing sender Actor to call RPC %s on target %s"), *Function->GetName(),
-				//	   *Actor->GetName());
-				// return;
+				SenderInfo.Entity = WorkerEntityId;
 			}
-			else
+			else if (!bHasSenderAvailable)
 			{
-				SenderActor = GSenderStack.Last();
-			}
+				if (bIsNetWriteFence)
+				{
+					UE_LOG(LogSpatialOSNetDriver, Error,
+						   TEXT("Net write fence %s on target %s will be dropped because no sender was provided"), *Function->GetName(),
+						   *Actor->GetName());
+					return;
+				}
+				else
+				{
+					UE_LOG(LogSpatialOSNetDriver, Warning,
+						   TEXT("Ordered reliable RPC %s on target %s will be sent unordered because no sender was provided"),
+						   *Function->GetName(), *Actor->GetName());
 
-			if (SenderActor == nullptr)
-			{
-				// These will eventually become error cases
-				// UE_LOG(LogSpatialOSNetDriver, Error, TEXT("Null sender Actor set to call RPC %s on target %s"), *Function->GetName(),
-				//	   *Actor->GetName());
-				// return;
+					SenderInfo.Entity = WorkerEntityId;
+				}
 			}
-
-			if (SenderActor && !SenderActor->HasAuthority())
+			else
 			{
-				UE_LOG(LogSpatialOSNetDriver, Error, TEXT("No authority on sender Actor %s to call RPC %s on target %s"),
-					   *SenderActor->GetName(), *Function->GetName(), *Actor->GetName());
-				return;
-			}
+				AActor* SenderActor = nullptr;
+
+				if (GSenderStack.Num() == 0)
+				{
+					UE_LOG(LogSpatialOSNetDriver, Error, TEXT("Missing sender Actor to call RPC %s on target %s"), *Function->GetName(),
+						   *Actor->GetName());
+					return;
+				}
+				else
+				{
+					SenderActorDesc& Desc = GSenderStack.Last();
+					SenderActor = Desc.Actor;
+
+					if (!ensure(!Desc.bHasBeenUsed))
+					{
+						UE_LOG(LogSpatialOSNetDriver, Error,
+							   TEXT("(INTERNAl GDK ERROR) Wrong stack semantics when calling RPC %s on Actor %s."
+									"The sender Actor is supposed to be used only once in a RPC case"),
+							   *Function->GetName(), *Actor->GetName());
+						return;
+					}
+
+					Desc.bHasBeenUsed = true;
+
+					if (bIsOnlyNetWriteFence && Desc.Kind != SenderActorDesc::Dependent
+						|| (!bIsNetWriteFence && Desc.Kind == SenderActorDesc::Dependent))
+					{
+						UE_LOG(
+							LogSpatialOSNetDriver, Error,
+							TEXT(
+								"Wrong kind of sender Actor set to call RPC %s on target %s."
+								"Check that the right AActor function was used with the right kind of RPC (CrossServer and NetWriteFence)"),
+							*Function->GetName(), *Actor->GetName());
+						return;
+					}
+				}
+
+				if (SenderActor == nullptr)
+				{
+					UE_LOG(LogSpatialOSNetDriver, Error, TEXT("Null sender Actor set to call RPC %s on target %s"), *Function->GetName(),
+						   *Actor->GetName());
+					return;
+				}
+
+				if (SenderActor && !SenderActor->HasAuthority())
```

Comment:
SendActor cannot be nullptr at this stage? Or did you mean to check whether actor is valid (not pending kill, etc)?

Reply:
I think this particular one was to allow some fallthrough, might be outdated.

Reply:
So can tidy up?